batch/batch/stoppable_server.py

```diff
@@ -0,0 +1,40 @@
+import asyncio
+import logging
+import sys
+
+from aiohttp import web
+
+
+class StoppableServer:
+    def __init__(self, app, host, port):
+        self.app = app
+        self.host = host
+        self.port = port
+        self.stop_queue = asyncio.Queue()
+        self.log = logging.getLogger('stoppable_server')
+
+    async def stop(self, exit_code):
+        await self.stop_queue.put(exit_code)
+
+    async def start(self):
+        runner = web.AppRunner(self.app)
+        await runner.setup()
+        site = web.TCPSite(runner, self.host, self.port)
+        self.log.info(f'serving {self.host}:{self.port}')
+        await site.start()
```

Comment:
Why does this need to be so complicated? Also would a semaphore be a better choice here?

Reply:
🤷‍♀ https://docs.aiohttp.org/en/stable/web_advanced.html#application-runners
If you want to stop the server, you need a runner. If you use a runner you need to create a site explicitly.

Hmm. We could use an `Event` or `Sempahore` both are simpler than a queue, but neither lets me communicate an exit code. I suppose I could put the exit code in a field and then notify the event, but the queue makes the value flow more clear.